mgmt/rest/plugin_v2.go

```diff
@@ -0,0 +1,411 @@
+		case control.ErrPluginNotInLoadedState.Error():
+			statusCode = 409
+		}
+		response.Write(statusCode, response.FromSnapError(se), w)
+		return
+	}
+	response.Write(204, nil, w)
+}
+
+func (s *Server) getPluginsV2(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+
+	// filter by plugin name or plugin type
+	plName := params.ByName("name")
+	plType := params.ByName("type")
+	nbFilter := Btoi(plName != "") + Btoi(plType != "")
+
+	if _, detail := r.URL.Query()["running"]; detail {
+		// get running plugins
+		plugins := runningPluginsBody(r.Host, s.mm.AvailablePlugins())
+		filteredPlugins := []response.RunningPlugin{}
+		if nbFilter > 0 {
+			for _, p := range plugins {
+				if nbFilter == 1 && (p.Name == plName || p.Type == plType) || nbFilter == 2 && (p.Name == plName && p.Type == plType) {
+					filteredPlugins = append(filteredPlugins, p)
+				}
+			}
+		} else {
+			filteredPlugins = plugins
+		}
+		response.Write(200, filteredPlugins, w)
+	} else  {
+		// get plugins from the plugin catalog
+		plugins := pluginCatalogBody(r.Host, s.mm.PluginCatalog())
+		filteredPlugins := []response.Plugin{}
+
+		if nbFilter > 0 {
+			for _, p := range plugins {
+				if nbFilter == 1 && (p.Name == plName || p.Type == plType) || nbFilter == 2 && (p.Name == plName && p.Type == plType) {
+					filteredPlugins = append(filteredPlugins, p)
+				}
+			}
+		} else {
+			filteredPlugins = plugins
+		}
+		response.Write(200, filteredPlugins, w)
+	}
+}
+
+func Btoi(b bool) int {
+	if b {
+		return 1
+	}
+	return 0
+}
+
+func pluginCatalogBody(host string, c []core.CatalogedPlugin) []response.Plugin {
+	plugins := make([]response.Plugin, len(c))
+	for i, p := range c {
+		plugins[i] = catalogedPluginBody(host, p)
+	}
+	return plugins
+}
+
+func catalogedPluginBody(host string, c core.CatalogedPlugin) response.Plugin {
+	return response.Plugin{
+		Name:            c.Name(),
+		Version:         c.Version(),
+		Type:            c.TypeName(),
+		Signed:          c.IsSigned(),
+		Status:          c.Status(),
+		LoadedTimestamp: c.LoadedTimestamp().Unix(),
+		Href:            pluginURI(host, "v2", c),
+	}
+}
+
+func runningPluginsBody(host string, c []core.AvailablePlugin) []response.RunningPlugin {
+	plugins := make([]response.RunningPlugin, len(c))
+	for i, p := range c {
+		plugins[i] = response.RunningPlugin{
+			Name:             p.Name(),
+			Version:          p.Version(),
+			Type:             p.TypeName(),
+			HitCount:         p.HitCount(),
+			LastHitTimestamp: p.LastHit().Unix(),
+			ID:               p.ID(),
+			Href:             pluginURI(host, "v2", p),
+			PprofPort:        p.Port(),
+		}
+	}
+	return plugins
+}
+
+func pluginURI(host, version string, c core.Plugin) string {
+	return fmt.Sprintf("%s://%s/%s/plugins/%s/%s/%d", protocolPrefix, host, version, c.TypeName(), c.Name(), c.Version())
+}
+
+func (s *Server) getPluginV2(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
+	plName := p.ByName("name")
```

Comment:
This section (lines 324-337) is repeated in unloadPlugin, should it be broken out into a separate method? If it's only those two places it doesn't have to be, but if it is in other places too it should be broken out.